artist.py

```diff
@@ -0,0 +1,72 @@
+import pygame
+import os
+from common_constants import Vector
+
+# in charge of drawing Game objects
+class Artist:
+    def __init__(self, game):
+        self.game = game
+
+        # pygame setup
+        os.environ['SDL_VIDEO_WINDOW_POS'] = "%d,%d" % (0,0) # locate the window in the upper left
+        pygame.init()
+        pygame.display.set_caption(f'{type(game).__name__} Reduction')
+
+        # compute constants
+
+        # this is the approximate height of the menu bar at the top of iOS
+        BUFFER = 28
+
+        # how much display space we have to work with
+        SCREEN_DIM = Vector(pygame.display.Info().current_w - BUFFER, pygame.display.Info().current_h - BUFFER)
+
+        # how big could each block dimension be if we used the whole display
+        FITTED_BLOCK_DIM = SCREEN_DIM / game.grid.dim
+
+        MIN_BLOCK_DIM = 15 # arbitrary cutoff for ease of visualization
+
+        # blocks need to be square and not smaller than the min
+        self.BLOCK_DIM = max(MIN_BLOCK_DIM, min(FITTED_BLOCK_DIM()))
+
+        # limit the max window dimension to fit on the display, and ensure it multiplies the block dimension
+        self.WINDOW_DIM = Vector(int(round(min(SCREEN_DIM.x - (SCREEN_DIM.x % self.BLOCK_DIM), self.BLOCK_DIM * game.grid.dim.x))), int(round(min(SCREEN_DIM.y - (SCREEN_DIM.y % self.BLOCK_DIM), self.BLOCK_DIM * game.grid.dim.y))))
+        
+        # the number of blocks that will fit in a window at a time
+        self.WINDOW_BLOCKS = Vector(int(round(self.WINDOW_DIM.x / self.BLOCK_DIM)), int(round(self.WINDOW_DIM.y / self.BLOCK_DIM)))
+        
+        # position of the player on the screen while scrolling is active
+        self.CENTER = Vector(int(self.WINDOW_BLOCKS.x / 2), int(self.WINDOW_BLOCKS.y / 2))
+
+        # store pygame resources
+        self.screen = pygame.display.set_mode(self.WINDOW_DIM())
+        self.clock = pygame.time.Clock()
+
+        # initial render
+        self.draw()
+
+    # all params are in block units
+    # return format is [corner x, corner y, width, height]
+    
+    # note that the y coordinate is inverted because pygame considers the upper
+    # left corner of the window to be the origin, while all logical objects
+    # rely on a more standard coordinate system originating in the lower left
+    def grid_to_px(self, x, y, offset):
+        return [(x - offset.x) * self.BLOCK_DIM + 1, (self.WINDOW_BLOCKS.y - 1 - (y - offset.y)) * self.BLOCK_DIM + 1, self.BLOCK_DIM - 1, self.BLOCK_DIM - 1]
+
+    def draw(self):
+        # compute the offset vector
+        offset = Vector(min(self.game.grid.dim.x - self.WINDOW_BLOCKS.x, max(0, self.game.player.pos.x - self.CENTER.x)), min(self.game.grid.dim.y - self.WINDOW_BLOCKS.y, max(0, self.game.player.pos.y - self.CENTER.y)))
```

Comment:
Same thing, I think it will be more readable separated (and you can see how the axes mirror each other which is nice)
```suggestion
        x_offset = min(self.game.grid.dim.x - self.WINDOW_BLOCKS.x, max(0, self.game.player.pos.x - self.CENTER.x))
        y_offset = min(self.game.grid.dim.y - self.WINDOW_BLOCKS.y, max(0, self.game.player.pos.y - self.CENTER.y))
        offset = Vector(x_offset, y_offset)
```